### PR TITLE
Verify variant uniqueness outside of Configuration

### PR DESCRIPTION
**Pull request by jvandort**
The preventFromFurtherMutation method did a lot more than the name implied. It also checked all other configurations to ensure they had separate attributes and capabilities than the current configuration being frozen. There are a few reasons we want to avoid doing this check in this method:

1. The configuration should not need a reference to the configurations provider. We are deprecating the getAll method because of this. This change is needed to avoid that reference
2. The method is more true to its name.

Furthermore, we create a FreezeableAttributeContainer which allows the attribute container to maintain the same attribute container reference before and after preventFromFurtherMutation is called. Previously, users who called getAttributes on the configuration before it was prevented from mutation would keep having the mutable reference after mutation was prevented. Now, those same callers will not be able to mutate their attribute container reference after the owning configuration is frozen.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle apt

**Comment by jvandort** [comment]
@bot-gradle test apt

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:717]: ❓ Why did this change to a call?

Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:717]: And line 822 and 1119

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1114]: 🎉

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentMetadata.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentMetadata.java:332]: 💭 You could switch this to return `Optional<GradleException>` and the use a more functional style chaining `.ifPresent()` to avoid returning null.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java:71]: 🤔 I think `IllegalStateException` is more appropriate here.  Also, putting at least the attribute name (possibly value?) in the error message would probably be helpful to the debugger who encounters this and tries to understand the behavior.

Why does this say "attributes of dependency X"?  No guarantee it's a dependency.  I think it would be better to be more generic "attributes associated with X"

Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java:71]: 🤔 Does this class need a public `isFrozen()` method?  Also the error checking here is an exact duplicate that could be extracted.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:111]: ❗  This only gets the first failure it encounters during iteration though.  If there are multiple different duplicate identities they won't be returned.  

Like, if configuration A == B, and C == D, then you have a multimap of (identity 1 -> A, B), (identity 2 -> C, D) and you only get A, B out of this.

Was this meant to show all collisions?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:73]: 💅 I think it's almost always worth making the APIs of new classes return empty `Optional`s instead of `null` for failure.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:88]: ❗ I think this find a matching configuration by name instead of instance.  Actually maybe the parameter should just be the name, too?  

I don't see why we should need to keep around the entire configuration to query the failures here, or why using a copy/clone with the same name shouldn't report the same failures.  At least while configuration live in a global container with globally unique names.

**Review by tresat — CHANGES_REQUESTED** [review]
Overall I think the change makes a lot of sense.  I left notes on a few small things to consider.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:717]: These were from another set of changes that this PR was split off from. They are not necessary for this change so I'll revert them. 

But for context, I was trying to deprecate incoming/outgoing so they are only used for the proper roles and I added the role guards to the `getIncoming`/`getOutgoing` methods

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentMetadata.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentMetadata.java:332]: In general I am not convinced that using `Optional` is any better than using `null`. Since we use null much more frequently throughout the codebase, I'd prefer to continue using that -- at least we should decide as a team to transition to using `Optional` more generally. 

In any case, I've updated the method to throw an exception instead of returning a value, so this doesn't matter here anymore

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java:71]: I don't think it needs a public `isFrozen` since at this point we never have the need for it. If we ever decide that we need to query the attribute container if its frozen, we could always add the method.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:111]: It was meant to do the same thing that the previous code did. The comment is misleading though. I've reworded it so that it no longer implies the exception encompasses all conflicts.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:73]: I don't necessarilly agree. Perhaps this is a good discussion to have parallel to this PR and I don't think we should block this change for this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java:71]: Not sure why the text was like this before. I've reworded the exception and updated it to `IllegalStateException`

Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java:71]: I've extracted the common part.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:88]: I've switched to a name comparison. However I've left the configuration as a parameter since we need to call `VariantIdentity.from(configuration)` and I'd rather not expose `VariantIdentity` right now

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java]: That's not to say that this message can't be improved, I just did not intend to spend the time improving it with this change

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/FreezableAttributeContainer.java]: I'll Let It Go then. ❄️

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java:73]: Yep, it definitely shouldn't block a change, just a nitpick.  I'll bring it up later to the larger group.

**Review by tresat — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**
